src/plugins/AvatarPEPPlugin.ts

```diff
@@ -0,0 +1,245 @@
+                           return [undefined];
+                        } else {
+                           return [avatar];
+                        }
+                     });
+               } else {
+                  return [avatar];
+               }
+            });
+      }
+   };
+
+   private onMessageAvatarUpdate = stanza => {
+      let from = new JID($(stanza).attr('from'));
+      let metadata = $(stanza).find('metadata[xmlns="urn:xmpp:avatar:metadata"]');
+      let data = $(stanza).find('data[xmlns="urn:xmpp:avatar:data"]');
+
+      if (metadata.length > 0) {
+         let info = metadata.find('info');
+         let contact = this.pluginAPI.getContact(from);
+         if (!contact) {
+            this.pluginAPI.Log.warn('No contact found for', from);
+            return true;
+         }
+
+         if (info.length > 0) {
+            let hash = $(info).attr('id');
+
+            let storedHash = this.getStorage().getItem(from.bare);
+            if (storedHash === undefined || hash !== storedHash) {
+               let avatarUI = AvatarUI.get(contact);
+               this.getStorage().setItem(contact.getJid().bare, hash);
+               avatarUI.reload();
+            }
+         } else {
+            let avatarUI = AvatarUI.get(contact);
+            this.getStorage().setItem(contact.getJid().bare, '');
+            avatarUI.reload();
+         }
+      } else if (data.length > 0) {
+         let contact = this.pluginAPI.getContact(from);
+         if (!contact) {
+            this.pluginAPI.Log.warn('No contact found for', from);
+            return true;
+         }
+
+         let src = data.text().replace(/[\t\r\n\f]/gi, '');
+         const b64str = src.replace(/^.+;base64,/, '');
+         let hash = Hash.SHA1FromBase64(b64str);
+         let avatarUI = AvatarUI.get(contact);
+         this.getStorage().setItem(contact.getJid().bare, hash);
+         avatarUI.reload();
+      }
+
+      return true;
+   };
+
+   private avatarProcessor = async (contact: IContact, avatar: IAvatar): Promise<[IContact, IAvatar]> => {
+      let storage = this.getStorage();
+      let hash = storage.getItem(contact.getJid().bare);
+
+      if (!hash && !avatar) {
+         try {
+            const avatarObject = await this.getAvatar(contact);
+            const data = avatarObject.src.replace(/^.+;base64,/, '');
+            avatar = new Avatar(Hash.SHA1FromBase64(data), avatarObject.type, avatarObject.src);
+
+            this.getStorage().setItem(contact.getJid().bare, avatar.getHash() || '');
+
+            let avatarUI = AvatarUI.get(contact);
+            avatarUI.reload();
+         } catch (err) {
+            // we could not find any avatar
+         }
+      }
+
+      if (!hash || avatar) {
+         return [contact, avatar];
+      }
+
+      try {
+         avatar = new Avatar(hash);
+      } catch (err) {
+         try {
+            const avatarObject = await this.getAvatar(contact);
```

Comment:
If there is no avatar, `this.getAvatar` should trhow an error, so the following check is not needed. Makes the code a bit easier to read.